### PR TITLE
Upgrade Zip to use lazy properties

### PR DESCRIPTION
**Pull request by cobexer**
### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/jvm/platform-jvm/src/main/java/org/gradle/jvm/tasks/Jar.java [hunk 1]
```diff
@@ -109,23 +111,8 @@ private OutputChangeListener outputChangeListener() {
...
-     *
-     * @param metadataCharset the character set used to encode JAR metadata like file names
-     * @since 2.14
-     */
-    @Override
-    public void setMetadataCharset(String metadataCharset) {
-        super.setMetadataCharset(metadataCharset);
-    }
+    @ReplacesEagerProperty
+    public abstract Property<String> getMetadataCharset();
```
Comment (cobexer) [inline platforms/jvm/platform-jvm/src/main/java/org/gradle/jvm/tasks/Jar.java:115]: ❓ I'm not sure this works as intended. This method only exists to change Javadoc AFAICT, is here a better way?

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/Zip.java [hunk 2]
```diff
@@ -109,39 +93,16 @@ public void setZip64(boolean allowZip64) {
      * This means you should not enable this property if you are building JARs to be used with Java 6 and earlier runtimes.
      */
     @Input
-    @ToBeReplacedByLazyProperty
-    public boolean isZip64() {
-        return allowZip64;
-    }
+    @ReplacesEagerProperty(originalType = boolean.class)
+    public abstract Property<Boolean> getZip64();
```
Comment (cobexer) [inline subprojects/core/src/main/java/org/gradle/api/tasks/bundling/Zip.java:97]: There is a smoke test failure:
>Execution failed for task ':shadowJar'.
  > Could not find matching constructor for: com.github.jengelman.gradle.plugins.shadow.internal.DefaultZipCompressor(org.gradle.api.internal.provider.DefaultProperty, Integer)


https://github.com/GradleUp/shadow/blob/b35e6904a1dfa541fc147d46e9ba456fdbe3f176/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/GradleVersionUtil.groovy#L18
```groovy
                return new DefaultZipCompressor(jar.zip64, ZipOutputStream.DEFLATED)
```

**Comment by cobexer** [comment]
@bot-gradle test this

File: platforms/jvm/platform-jvm/src/main/java/org/gradle/jvm/tasks/Jar.java (hunk 1 above)
Comment (cobexer) [reply platforms/jvm/platform-jvm/src/main/java/org/gradle/jvm/tasks/Jar.java:115]: Removing `@ReplacesEagerProperty` so the update machinery doesn't attempt anything here.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test RFN

**Comment by asodja** [comment]
@bot-gradle test this

**Pull request merged**